lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,502 @@
+
+// Lower instance operations in the same way as module block arguments. Bundles
+// are flattened, and other arguments are copied to keep the relative order. By
+// ensuring both lowerings are the same, we can process every module in the
+// circuit in parallel, and every instance will have the correct ports.
+LogicalResult FIRRTLTypesLowering::visitDecl(InstanceOp op) {
+  auto originalType = op.result().getType().dyn_cast<BundleType>();
+  if (!originalType) {
+    op.emitError("instance result was not bundle type");
+    return failure();
+  }
+
+  // Create a new, flat bundle type for the new result
+  SmallVector<BundleType::BundleElement, 8> bundleElements;
+  for (auto element : originalType.getElements()) {
+    if (element.second.isa<BundleType>()) {
+      // Flatten nested bundle types the usual way.
+      SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+      flattenBundleTypes(element.second, element.first, false, fieldTypes);
+
+      for (auto field : fieldTypes) {
+        // Store the flat type for the new bundle type.
+        auto flatName = builder->getIdentifier(field.suffix);
+        auto flatType = getPortType(field);
+        auto newElement = BundleType::BundleElement(flatName, flatType);
+        bundleElements.push_back(newElement);
+      }
+    } else {
+      // Store the original type for the new bundle type.
+      bundleElements.push_back(element);
+    }
+  }
+
+  // Get the new bundle type and create a new instance.
+  auto newType = BundleType::get(bundleElements, &getContext());
+
+  auto newInstance = builder->create<InstanceOp>(
+      op.getLoc(), newType, op.moduleNameAttr(), op.nameAttr());
+
+  // Store the mapping from the old instance to the new one.
+  setInstanceLowering(op, newInstance);
+
+  removeOp(op);
+
+  return success();
+}
+
+// Lowering subfield operations has to deal with three different cases:
+//   1. the input value is from a module block argument
+//   2. the input value is from another subfield operation's result
+//   3. the input value is from an instance
+LogicalResult FIRRTLTypesLowering::visitExpr(SubfieldOp op) {
+  Value input = op.input();
+  Value result = op.result();
+  StringRef fieldname = op.fieldname();
+
+  Value rootBundle;
+  SmallString<16> suffix;
+  if (input.isa<BlockArgument>()) {
+    // When the input is from a block argument, map the result to the root
+    // bundle block argument, and begin creating a suffix string.
+    rootBundle = input;
+    suffix.push_back('_');
+    suffix += fieldname;
+  } else if (input.isa<OpResult>()) {
+    Operation *owner = input.cast<OpResult>().getOwner();
+    if (isa<SubfieldOp>(owner)) {
+      // When the input is from a subfield, look up the lowered info from the
+      // parent.
+      auto subfieldLowering = getSubfieldLowering(input);
+      if (!subfieldLowering.hasValue()) {
+        op.emitError("didn't find subfield lowering for input");
```

Comment:
Now that I think about it, all of these errors are more like `assert`s. The pass is happy to leave everything but bundles alone right now. I think we can only get into these situations if the pass itself has mishandled an otherwise valid bundle.

Reply:
:+1: for making these asserts if there are no cases where the transform would fail to find a lowering (except for invalid input, which I agree should go into the verifier).

Reply:
I'm going through these and updating them to asserts where I think it makes sense. If it's possible to hit one of these cases with verified IR, I'll add a test for the expected error in that scenario.